docs/zh/config-desc.md
# 配置文件

默认配置文件是～/.sophon-messager/config.toml。`#`表示是注释。

```toml
[api]
  Address = "/ip4/127.0.0.1/tcp/39812"  #messager的监听地址

[db]
  type = "sqlite"  #数据库类型。mysql或者sqlite

  [db.mysql]
    connMaxLifeTime = "1m0s"
    connectionString = ""
    debug = false
    maxIdleConn = 10
    maxOpenConn = 10

  [db.sqlite]
    debug = false

[gateway]
  token = ""   #[gateway],[jwt],[node]三个字段基本上都是用同一个auth服务的token
  url = ["/ip4/127.0.0.1/tcp/45132"]

[jwt]
  authURL = "http://127.0.0.1:8989"
  token = "" #[gateway],[jwt],[node]三个字段基本上都是用同一个auth服务的token

# messager直接通过p2p给链节点（venus/lotus）发送消息
# 可选
[libp2p]
  bootstrapAddresses = []
  expandPeriod = "0s"
  listenAddresses = "/ip4/0.0.0.0/tcp/0"
  minPeerThreshold = 0

[messageService]
  DefaultTimeout = "1s"  #请求链节点接口的超时时长 
  EstimateMessageTimeout = "5s" #调用链节点进行消息预估gas费等的超时时长
  SignMessageTimeout = "3s" #调用gateway请求wallet进行签名的超时时长
  WaitingChainHeadStableDuration = "8s" #messager收到一个newhead消息后，如果8秒内没有收到新的newhead，就会认为收到的newhead是stable的了
  skipProcessHead = false #是否更新消息上链后的状态。在多个messager共用一个数据库时，只需要一个messager进行消息的全部状态更新
  skipPushMessage = false  #不推送消息到链。在多个messager共用一个数据库时，不推送消息的messager只做接受消息的任务，另外的messager进行推送消息

[metrics]
  Enabled = false

  [metrics.Exporter]
    Type = "prometheus"

    [metrics.Exporter.Graphite]
      Host = "127.0.0.1"
      Namespace = ""
      Port = 4568
      ReportingPeriod = "10s"

    [metrics.Exporter.Prometheus]
      EndPoint = "/ip4/0.0.0.0/tcp/4568"
      Namespace = ""
      Path = "/debug/metrics"
      RegistryType = "define"
      ReportingPeriod = "10s"

#venus同步节点或者sophon-co负载代理服务
[node]
  token = "" #[gateway],[jwt],[node]三个字段基本上都是用同一个auth服务的token
  url = "/ip4/127.0.0.1/tcp/3453"

[publisher]
  cacheReleasePeriod = 0 #间隔多久缓存清理一次
  concurrency = 5 #同时推送消息的线程数
  enableMultiNode = true #是否可以给多个节点推送消息
  enablePubsub = false #是否通过p2p网络发送消息。需要和[libp2p]配置一起使用

#可选
[rateLimit]
  #redis地址，用于记录用户访问的次数。如果要开启对某个user的访问限速，还需要`auth` 服务同时设置`sophon-auth user rate-limit`命令。
  redis = ""  # eg. 127.0.0.1:6379

#可选
[tracing]
  JaegerEndpoint = "" # 例如：localhost:6831
  JaegerTracingEnabled = false
  ProbabilitySampler = 1.0
  ServerName = ""

```

docs/zh/metrics-config-desc.md
# metrics 配置及使用说明

## 配置

`Metrics` 基本的配置样例如下：
```toml
[Metrics]
  # 是否开启metrics指标统计，默认为false
  Enabled = false
  
  [Metrics.Exporter]
    # 指标导出器类型，目前可选：prometheus或graphite，默认为prometheus
    Type = "prometheus"
    
    [Metrics.Exporter.Prometheus]
      # multiaddr
      EndPoint = "/ip4/0.0.0.0/tcp/4568"
      # 命名规范: "a_b_c", 不能带"-"
      Namespace = "messager01" 
      # 指标注册表类型，可选：default（默认，会附带程序运行的环境指标）或 define（自定义）
      RegistryType = "define"
      # prometheus 服务路径
      Path = "/debug/metrics"
      # 上报周期
      ReportingPeriod = "10s"
      
    [Metrics.Exporter.Graphite]
      # 命名规范: "a_b_c", 不能带"-"
      Namespace = "messager01" 
      # graphite exporter 收集器服务地址
      Host = "127.0.0.1"
      # graphite exporter 收集器服务监听端口
      Port = 4568
      # 上报周期
      ReportingPeriod = "10s"
```
## 导出器

目前可以选择两类导出器（`exporter`）：`Prometheus exporter` 或 `Graphite exporter`，默认是前者。

如果配置 `Prometheus exporter`，则在 `venus-messager` 服务启动时会附带启动 `Prometheus exporter` 的监听服务，可以通过以下方式快速查看指标：


```bash
 $  curl http://localhost:4568/debug/metrics
   # HELP messager01_chain_head_stable_dur_s Duration of chain head stabilization
   # TYPE messager01_chain_head_stable_dur_s histogram
   messager01_chain_head_stable_dur_s_bucket{le="8"} 0
   messager01_chain_head_stable_dur_s_bucket{le="9"} 11
   messager01_chain_head_stable_dur_s_bucket{le="10"} 27
   messager01_chain_head_stable_dur_s_bucket{le="12"} 43
   messager01_chain_head_stable_dur_s_bucket{le="14"} 48
   messager01_chain_head_stable_dur_s_bucket{le="16"} 49
   messager01_chain_head_stable_dur_s_bucket{le="18"} 49
   messager01_chain_head_stable_dur_s_bucket{le="20"} 49
   messager01_chain_head_stable_dur_s_bucket{le="25"} 49
   messager01_chain_head_stable_dur_s_bucket{le="30"} 49
   messager01_chain_head_stable_dur_s_bucket{le="60"} 50
   messager01_chain_head_stable_dur_s_bucket{le="+Inf"} 50
   messager01_chain_head_stable_dur_s_sum 503.99999999999983
   messager01_chain_head_stable_dur_s_count 50
   # HELP messager01_chain_head_stable_s Delay of chain head stabilization
   # TYPE messager01_chain_head_stable_s gauge
   messager01_chain_head_stable_s 9
   ... ...
```
> 如果遇到错误 `curl: (56) Recv failure: Connection reset by peer`, 请使用本机 `ip` 地址, 如下所示:
```bash
$  curl http://<ip>:4568/debug/metrics
```

如果配置 `Graphite exporter`，需要先启动 `Graphite exporter` 的收集器服务， `venus-messager` 服务启动时将指标上报给收集器。服务启动参考 [Graphite exporter](https://github.com/prometheus/graphite_exporter) 中的说明。

`Graphite exporter` 和 `Prometheus exporter` 自身都不带图形界面的，如果需要可视化监控及更高阶的图表分析，请到 `venus-docs` 项目中查找关于 `Prometheus+Grafana` 的说明文档。


CHANGELOG.md
# sophon-message changelog

## v1.12.0

* fix: add repo compatibility for cli [[#350](https://github.com/ipfs-force-community/sophon-messager/pull/350)]

## v1.12.0-rc1

* fix: repeat process msg [[#340](https://github.com/ipfs-force-community/sophon-messager/pull/340)]
* opt: clear invalid error msg [[#343](https://github.com/ipfs-force-community/sophon-messager/pull/343)]
* feat: add HasActorCfg method [[#344](https://github.com/ipfs-force-community/sophon-messager/pull/344)]
* chore: upgrade venus and venus-auth [[#345](https://github.com/ipfs-force-community/sophon-messager/pull/345)]
* feat: rename project [[#346](https://github.com/ipfs-force-community/sophon-messager/pull/346)]

## v1.11.0

* bump up version to v1.11.0

## v1.11.0-rc1

### Features
* feat: support method type fee cfg / 支持消息类型级别的费用配置  [[#303](https://github.com/ipfs-force-community/sophon-messager/pull/303)]
* feat: add status api to detect api ready p /添加状态检测接口 [[#313](https://github.com/ipfs-force-community/sophon-messager/pull/313)]
* feat: update the authClient with token  /客户端token验证 [[#317](https://github.com/ipfs-force-community/sophon-messager/pull/317)]
* chore: more detailed error information /更加详细的错误信息 [[#331](https://github.com/ipfs-force-community/sophon-messager/pull/331)]
* feat: ListBlockedMessage interface also returns unfill message  / 同样返回Unfill的消息 [[#330](https://github.com/ipfs-force-community/sophon-messager/pull/330)]
* feat: add docker push  / 增加推送到镜像仓库的功能 [[#335](https://github.com/ipfs-force-community/sophon-messager/pull/335)]
* feat: Reduce minimum replace fee from 1.25x to 1.1x  / 最小 replace fee 乘数改为 1.1x [[#336](https://github.com/ipfs-force-community/sophon-messager/pull/336)]


### Bug Fixes
* fix: No actor configuration was used when replacing messages  /replace 消息时没有使用 actor config 表中的 maxfee[[#328](https://github.com/ipfs-force-community/sophon-messager/pull/328)]
* fix: failed create actor_cfg table in mysql  / 修复创建actor_cfg表失败[[#327](https://github.com/ipfs-force-community/sophon-messager/pull/327)]
* fix: Exclude empty strings when listing failure message /排除 error_msg 为空的时候被认为是 failed 消息[[#329](https://github.com/ipfs-force-community/sophon-messager/pull/329)]
* fix: Modify the WaitingChainHeadStableDuration value only for 2k networks  / 2k网络中修改WaitingChainHeadStableDuration的值[[#334](https://github.com/ipfs-force-community/sophon-messager/pull/334)]

## v1.10.1

* 支持 delegated 地址的消息 [[#323](https://github.com/ipfs-force-community/sophon-messager/pull/323)]
* 升级 venus 和 venus-auth 版本到 v1.10.1

## v1.10.0

* 升级 venus 和 venus-auth 版本到 v1.10.0
* 升级 go-jsonrpc 版本到 v0.1.7

## v1.10.0-rc3

1. 升级 venus 和 go-jsonrpc 版本
2. 推送消息接口先解析地址再验证权限

## v1.10.0-rc1

支持 Filecoin NV18 网络升级

* 把 replacedmsg 重命名为 nonceconfict [[#306](https://github.com/ipfs-force-community/sophon-messager/pull/304)]
* 使用 untrust 接口推送消息 [[#306](https://github.com/ipfs-force-community/sophon-messager/pull/306)]
* 按从小到大查询 unchain 消息 [[#307](https://github.com/ipfs-force-community/sophon-messager/pull/307)]
* 移除测试中的重复代码 [[#308](https://github.com/ipfs-force-community/sophon-messager/pull/308)]


config/config.go
package config

import (
	"time"

	"github.com/ipfs-force-community/metrics"
)

type Config struct {
	DB             DbConfig               `toml:"db"`
	JWT            JWTConfig              `toml:"jwt"`
	Log            LogConfig              `toml:"log"`
	API            APIConfig              `toml:"api"`
	Node           NodeConfig             `toml:"node"`
	MessageService MessageServiceConfig   `toml:"messageService"`
	Gateway        GatewayConfig          `toml:"gateway"`
	RateLimit      RateLimitConfig        `toml:"rateLimit"`
	Trace          *metrics.TraceConfig   `toml:"tracing"`
	Metrics        *metrics.MetricsConfig `toml:"metrics"`
	Libp2pNet      *Libp2pNetConfig       `toml:"libp2p"`
	Publisher      *PublisherConfig       `toml:"publisher"`
}

type NodeConfig struct {
	Url   string `toml:"url"`
	Token string `toml:"token"`
}

type LogConfig struct {
	Path  string `toml:"path"`
	Level string `toml:"level"`
}

type APIConfig struct {
	Address string
}

type DbConfig struct {
	Type   string       `toml:"type"`
	MySql  MySqlConfig  `toml:"mysql"`
	Sqlite SqliteConfig `toml:"sqlite"`
}

type SqliteConfig struct {
	Debug bool `toml:"debug"`
}

type MySqlConfig struct {
	ConnectionString string        `toml:"connectionString"`
	MaxOpenConn      int           `toml:"maxOpenConn"`
	MaxIdleConn      int           `toml:"maxIdleConn"`
	ConnMaxLifeTime  time.Duration `toml:"connMaxLifeTime"`
	Debug            bool          `toml:"debug"`
}

type JWTConfig struct {
	AuthURL string `toml:"authURL"`
	Token   string `toml:"token"`
}

const (
	MinWaitingChainHeadStableDuration = time.Second * 2
	MaxWaitingChainHeadStableDuration = time.Second * 25
	DefWaitingChainHeadStableDuration = time.Second * 8
)

const (
	DefaultTimeout         = time.Second
	SignMessageTimeout     = time.Second * 3
	EstimateMessageTimeout = time.Second * 5
)

type MessageServiceConfig struct {
	WaitingChainHeadStableDuration time.Duration `toml:"WaitingChainHeadStableDuration"`

	DefaultTimeout         time.Duration `toml:"DefaultTimeout"`
	SignMessageTimeout     time.Duration `toml:"SignMessageTimeout"`
	EstimateMessageTimeout time.Duration `toml:"EstimateMessageTimeout"`

	SkipProcessHead bool `toml:"skipProcessHead"`
	SkipPushMessage bool `toml:"skipPushMessage"`
}

type Libp2pNetConfig struct {
	ListenAddress      string   `toml:"listenAddresses"`
	BootstrapAddresses []string `toml:"bootstrapAddresses"`

	// MinPeerThreshold determine when to expand peers.
	// default set to 0 which means use network default config.
	MinPeerThreshold int `toml:"minPeerThreshold"`

	// ExpandPeriod determine how often to expand peers.
	// default set to "0s" which means use network default config.
	// otherwise, it should be a duration string like "5s", "30s".
	ExpandPeriod time.Duration `toml:"expandPeriod"`
	// TODO: EnableRelay
}

type PublisherConfig struct {
	// CacheReleasePeriod is the period to release massage cache with unit of second.
	// default is 5.
	// set a negative int means disable cache.
	Concurrency int `toml:"concurrency"`

	// CacheReleasePeriod is the period to release massage cache with unit of second.
	// default is 0 which means auto decide by network parameters that is 1/3 of the block time.
	// set a negative int means disable cache
	CacheReleasePeriod int64 `toml:"cacheReleasePeriod"`

	EnableP2P       bool `toml:"enablePubsub"`
	EnableMultiNode bool `toml:"enableMultiNode"`
}

type MessageStateConfig struct {
	BackTime int `toml:"backTime"` // 向前找多久的数据写到内存,单位秒

	DefaultExpiration, CleanupInterval int // message 缓存的有效时间和清理间隔
}

type GatewayConfig struct {
	Token string   `toml:"token"`
	Url   []string `toml:"url"`
}

type RateLimitConfig struct {
	Redis string `toml:"redis"`
}

func DefaultConfig() *Config {
	return &Config{
		DB: DbConfig{
			Type: "sqlite",
			MySql: MySqlConfig{
				ConnectionString: "",
				MaxOpenConn:      10,
				MaxIdleConn:      10,
				ConnMaxLifeTime:  time.Second * 60,
				Debug:            false,
			},
		},
		JWT: JWTConfig{
			AuthURL: "http://127.0.0.1:8989",
		},
		Log: LogConfig{
			Path:  "",
			Level: "info",
		},
		API: APIConfig{
			Address: "/ip4/127.0.0.1/tcp/39812",
		},
		Node: NodeConfig{
			Url:   "/ip4/127.0.0.1/tcp/3453",
			Token: "",
		},
		MessageService: MessageServiceConfig{
			WaitingChainHeadStableDuration: DefWaitingChainHeadStableDuration,

			DefaultTimeout:         DefaultTimeout,
			SignMessageTimeout:     SignMessageTimeout,
			EstimateMessageTimeout: EstimateMessageTimeout,

			SkipProcessHead: false,
			SkipPushMessage: false,
		},
		Gateway: GatewayConfig{
			Token: "",
			Url:   []string{"/ip4/127.0.0.1/tcp/45132"},
		},
		RateLimit: RateLimitConfig{Redis: ""},
		Trace:     metrics.DefaultTraceConfig(),
		Metrics:   metrics.DefaultMetricsConfig(),
		Libp2pNet: &Libp2pNetConfig{
			ListenAddress:      "/ip4/0.0.0.0/tcp/0",
			BootstrapAddresses: []string{},
			MinPeerThreshold:   0,
			ExpandPeriod:       0 * time.Second,
		},
		Publisher: &PublisherConfig{
			Concurrency:        5,
			CacheReleasePeriod: 0,
			EnableP2P:          false,
			EnableMultiNode:    true,
		},
	}
}
